contracts/RestrictedNORI.sol

```diff
@@ -0,0 +1,1034 @@
+   * - The requirements of _beforeTokenTransfer apply to this function
+   * - toAccounts.length == removalIds.length == amounts.length
+   */
+  function batchRevokeUnreleasedTokenAmounts(
+    address[] calldata toAccounts,
+    uint256[] calldata removalIds,
+    uint256[] calldata amounts
+  ) external whenNotPaused onlyRole(TOKEN_REVOKER_ROLE) {
+    if (!(toAccounts.length == removalIds.length)) {
+      revert ArrayLengthMismatch({
+        array1Name: "fromAccounts",
+        array2Name: "removalIds"
+      });
+    }
+    if (!(toAccounts.length == amounts.length)) {
+      revert ArrayLengthMismatch({
+        array1Name: "fromAccounts",
+        array2Name: "amounts"
+      });
+    }
+    for (uint256 i = 0; i < toAccounts.length; i++) {
+      _revokeUnreleasedTokens(toAccounts[i], removalIds[i], amounts[i]);
+    }
+  }
+
+  // Private implementations ==========================================
+
+  /**
+   * Wraps minting of wrapper token and schedule setup.
+   *
+   * @dev If no schedule is set up for the specified removal id, one is created.
+   *
+   * @param removalId uint256 The removal for which funds are being deposited.
+   * @param amount uint256 Quantity of `_bridgedPolygonNori` to deposit
+   */
+  function _depositFor(
+    uint256 removalId,
+    uint256 amount,
+    bytes memory,
+    bytes memory
+  ) internal returns (bool) {
+    uint256 projectId = _removal.getProjectIdForRemoval(removalId);
+    Removal.ScheduleData memory scheduleData = _removal
+      .getScheduleDataForProjectId(projectId);
+    address recipient = scheduleData.supplierAddress;
+    if (!_scheduleIdToScheduleStruct[projectId].exists) {
+      _createSchedule(projectId);
+    }
+    super._mint(recipient, projectId, amount, "");
+    Schedule storage schedule = _scheduleIdToScheduleStruct[projectId];
+    // slither-disable-next-line unused-return address may already be in set and that is ok
+    schedule.tokenHolders.add(recipient);
+    // slither-disable-next-line unused-return id may already be in the set and that is ok
+    _addressToScheduleIdSet[recipient].add(projectId);
+    return true;
+  }
+
+  /**
+   * Sets up a schedule for the specified project id (implementation).
+   *
+   * @dev schedules are created when removal tokens are listed for sale in the market contract,
+   * so this should only be invoked during `tokensReceived` in the exceptional case that
+   * tokens were sent to this contract without a schedule set up.
+   */
+  function _createSchedule(uint256 projectId) internal {
+    Removal.ScheduleData memory scheduleData = _removal
+      .getScheduleDataForProjectId(projectId);
+    if (scheduleData.startTime == 0) {
+      revert InvalidScheduleStartTime({projectId: projectId});
+    }
+    address recipient = scheduleData.supplierAddress;
+    if (recipient == address(0)) {
+      revert RecipientCannotBeZeroAddress();
+    }
+    if (hasRole(SCHEDULE_CREATOR_ROLE, recipient)) {
```

Comment:
why does the recipient need a `SCHEDULE_CREATOR_ROLE` ? is the recipient not the supplier?

Reply:
they don't... this is keeping anyone who is a schedule creator (such as nori admin or the market contract) from being able to set one up for themselves, but I think we can get rid of this because it's also a sort of relic from `LockedNORI` where it actually matters that the receiver of a grant can't also be the grant admin